include/pisa/query/queries.hpp

```diff
@@ -32,70 +26,112 @@ namespace pisa {
         auto raw_term = *term_iter;
         try {
-            auto processed = process_term(std::string(raw_term));
-            if (processed) {
-                if (not stopwords or stopwords->find(*processed) == stopwords->end()) {
-                    parsed_query.push_back(std::move(*processed));
+            auto term = term_processor(raw_term);
+            if (term) {
+                if (!term_processor.is_stopword(*term)) {
+                    parsed_query.push_back(std::move(*term));
                 } else {
-                    spdlog::warn("Term `{}` is a stop word and will be ignored", *processed);
+                    spdlog::warn("Term `{}` is a stopword and will be ignored", *term);
                 }
             } else {
                 spdlog::warn("Term `{}` not found and will be ignored", raw_term);
             }
-        } catch (std::invalid_argument& err) {
+        } catch (std::invalid_argument &err) {
             spdlog::warn("Could not parse `{}` to a number", raw_term);
         }
     }
-    return {id, parsed_query, {}};
+    return {std::move(id_and_raw_query.first), std::move(parsed_query), {}};
+}
+
+[[nodiscard]] auto parse_query_ids(std::string const &query_string) -> Query
+{
+    auto id_and_raw_query = split_query_at_colon(query_string);
+    auto id = id_and_raw_query.first;
+    auto raw_query = id_and_raw_query.second;
+
+    std::vector<term_id_type> parsed_query;
+    std::vector<std::string> term_ids;
+    boost::split(term_ids, raw_query, boost::is_any_of("\t"));
+    try {
+        std::transform(term_ids.begin(),
+                       term_ids.end(),
+                       std::back_inserter(parsed_query),
+                       [](const std::string &val) { return std::stoi(val); });
+    } catch (std::invalid_argument &err) {
+        spdlog::error("Could not parse term identifiers of query `{}`", raw_query);
+        exit(1);
+    }
+    return {std::move(id), std::move(parsed_query), {}};
 }
 
-bool read_query(term_id_vec &ret, std::istream &is = std::cin,
-                std::function<term_id_type(std::string)> process_term = [](auto str) {
-                    return std::stoi(str);
-                }) {
+[[nodiscard]] std::function<void(const std::string)> compute_parse_query_function(
+    std::vector<Query> &queries,
+    const std::optional<std::string> terms_file,
+    const std::optional<std::string> stopwords_filename,
+    const std::optional<std::string> stemmer_type) {
+    if (terms_file) {
+        auto term_processor = TermProcessor(terms_file, stopwords_filename, stemmer_type);
+        return [&](std::string const &query_line) {
+            queries.push_back(parse_query_terms(query_line, term_processor));
+        };
+    } else {
+        return
+            [&](std::string const &query_line) { queries.push_back(parse_query_ids(query_line)); };
+    }
+}
+
+bool read_query(term_id_vec &ret, std::istream &is = std::cin)
```

Comment:
So here this is confusing, I don't see any reason why such a function should exist but not support string terms; that's weird. Are there many places we use this one? If not, maybe we can get rid of it altogether---my feeling is that parsing a query line (like the function above does) should be decoupled from reading actual lines. This can be done by any number of ways, but most notably there is a `io::for_each_line` function that is used in many places anyway. I think this is a better approach.

Reply:
I check this function isn't used anywhere (instead it is used `io::for_each_line`). Thus, probably it must be removed.

Reply:
Yeah, if not used, let's remove it. It's legacy code.

Reply:
@elshize, I was wrong. This function is used in `selective_queries` and `profile_queries`.

Reply:
Ok, that's cool, let's just leave it. I'll create a separate issue for that so we don't forget, we can take any relevant discussion there.